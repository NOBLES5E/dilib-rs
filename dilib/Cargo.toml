[package]
name = "dilib"
description = "A dependency injection library for Rust"
keywords = ["rust-patterns"]
version = "0.2.0-alpha"
authors = ["Freddy L."]
edition = "2021"
license-file = "LICENSE"
repository = "https://github.com/Neo-Ciber94/dilib-rs/"
homepage = "https://crates.io/crates/dilib/"
documentation = "https://docs.rs/dilib/"
rust-version = "1.59.0"

[package.metadata.docs.rs]
all-features = true
rustdoc-args = ["--cfg", "doc_cfg"]

[features]
# Nothing by default
default=[]

# All features excluding unstables
full = ["derive", "global", "lazy"]

# Enables #[derive(Inject)] proc macro
derive = ["dilib-derive"]

# Enables global `Container`
global = []

# Enables lazy singletons
lazy = ["once_cell"]

# Enables #[provide(...)] proc macro for global `Container`.
# This feature is considered unstable because `Rust` does not guarantee
# that code will always be allowed to executed before `main`
# until global constructor get implemented, if ever happen.
unstable_provide = ["dilib-macros", "global", "ctor", "lazy"]

[dependencies]
dilib-derive = { path="../dilib-derive", optional = true }
dilib-macros = { path = "../dilib-macros", optional = true}
ctor = { version ="0.1.22", optional = true }
once_cell = { version = "1.10.0", optional = true }

[dev-dependencies]
dilib-derive = { path="../dilib-derive" }
dilib = { path = ".", features=["full", "unstable_provide"]}