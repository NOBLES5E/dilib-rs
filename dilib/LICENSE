D:/Coding/Projects/Rust/dilib-rs/LICENSE