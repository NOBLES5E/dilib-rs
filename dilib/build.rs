use std::fs;
use std::path::PathBuf;

fn main() {
    println!("cargo:rerun-if-changed=README.md");

    let readme = fs::read_to_string("README.md").expect("Failed to read README.md");
    let output = PathBuf::from(std::env::var("OUT_DIR").unwrap()).join("README-lib.md");
    fs::write(output, readme).expect("Failed to write README-lib.md");
}
