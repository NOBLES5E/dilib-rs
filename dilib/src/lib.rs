//! # dilib-rs
//!
//! A dependency injection library for Rust.
//!
//! ## Usage
//! ```toml
//! [dependencies]
//! dilib = "0.1.3-alpha"
//! ```
//!
//! ## Example
//!
//! ### Basic Usage
//!
//! ```rust
//! use std::sync::Mutex;
//! use dilib::*;
//!
//! struct GetNumber(u32);
//!
//! struct Counter {
//!     val: u32,
//! }
//!
//! let mut container = Container::new();
//! container.add_scoped(|| String::from("Hello, world!"));
//! container.add_scoped_with_name("42", || GetNumber(42));
//! container.add_scoped_with_name("102", || GetNumber(102));
//! container.add_singleton(Mutex::new(Counter { val: 0 }));
//!
//! let mut counter = container.get_singleton::<Mutex<Counter>>().unwrap();
//! counter.lock().unwrap().val = 5;
//!
//! assert_eq!(container.get_scoped::<String>().unwrap(), "Hello, world!".to_string());
//! assert_eq!(container.get_scoped_with_name::<GetNumber>("42").unwrap().0, 42);
//! assert_eq!(container.get_scoped_with_name::<GetNumber>("102").unwrap().0, 102);
//! assert_eq!(container.get_singleton::<Mutex<Counter>>().unwrap().lock().unwrap().val, 5);
//! ```
//!
//! ### With derive
//! Requires `derive` feature.
//!
//! ```rust
//! use dilib::*;
//! use std::sync::Mutex;
//!
//! #[derive(Injectable)]
//! struct Counter {
//!     #[inject(default="Counter")]
//!     label: String,
//!     val: Singleton<Mutex<usize>>
//! }
//!
//! let mut container = Container::new();
//! container.add_singleton(Mutex::new(0_usize));
//! container.add_deps::<Counter>();
//!
//! let c1 = container.get_scoped::<Counter>().unwrap();
//! *c1.val.lock().unwrap() = 12;
//!
//! let c2 = container.get_scoped::<Counter>().unwrap();
//! assert_eq!(c1.label, c2.label);
//! assert_eq!(*c2.val.lock().unwrap(), 12);
//! ```

mod lib_macros;
pub use lib_macros::*;

mod injectable;
pub use injectable::*;

mod key;
pub use key::*;

mod container;
pub use container::*;

mod provider;
pub use provider::*;

mod scoped;
pub use scoped::*;

mod shared;
pub use shared::*;

mod resolved;
pub use resolved::*;

#[cfg(feature = "derive")]
extern crate dilib_derive;

#[cfg(feature = "derive")]
pub use dilib_derive::*;

/// A global instance of `Container`.
#[cfg(feature = "global")]
pub mod global;

/// procedural macros of `dilib`.
#[cfg(feature = "unstable_provide")]
pub mod macros {
    pub use dilib_macros::*;
}
