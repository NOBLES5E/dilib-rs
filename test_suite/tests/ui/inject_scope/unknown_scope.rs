use dilib::{Singleton, Injectable};

#[derive(Injectable)]
struct MyStruct {
    #[inject(scope="unique")]
    a: Singleton<usize>,
}

fn main(){}