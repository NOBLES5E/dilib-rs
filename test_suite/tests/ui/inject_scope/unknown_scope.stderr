error: proc-macro derive panicked
 --> $DIR/unknown_scope.rs:3:10
  |
3 | #[derive(Injectable)]
  |          ^^^^^^^^^^
  |
  = help: message: invalid scope value: `unique`, expected "singleton" or "scoped"
